File: mvvm/src/main/java/com/loror/mvvm/adapter/HolderBaseAdapter.java
package com.loror.mvvm.adapter;

import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import androidx.annotation.NonNull;

import java.util.List;

public abstract class HolderBaseAdapter<T, VH extends HolderBaseAdapter.ViewHolder> extends BaseAdapter {

    private final List<T> data;

    public HolderBaseAdapter(List<T> data) {
        this.data = data;
    }

    @Override
    public int getCount() {
        return data == null ? 0 : data.size();
    }

    @Override
    public T getItem(int position) {
        return data == null ? null : data.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        VH viewHolder;
        if (convertView == null) {
            viewHolder = onCreateViewHolder(parent, getItemViewType(position));
            convertView = viewHolder.itemView;
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (VH) convertView.getTag();
        }
        onBindViewHolder(viewHolder, position);
        return convertView;
    }

    public abstract VH onCreateViewHolder(ViewGroup parent, int viewType);

    public abstract void onBindViewHolder(@NonNull VH holder, int position);

    public static class ViewHolder {

        protected View itemView;

        public ViewHolder(@NonNull View itemView) {
            if (itemView == null) {
                throw new IllegalArgumentException("itemView may not be null");
            }
            this.itemView = itemView;
        }
    }
}

File: app/src/main/java/com/loror/mvvmbase/util/HolderBaseAdapter.kt
package com.loror.mvvmbase.util

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding

/**
 * Created by Loror on 2017/7/6.
 */
abstract class HolderBaseAdapter(val context: Context) : BaseAdapter() {
    val inflater: LayoutInflater
    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    @LayoutRes
    abstract fun getLayout(position: Int): Int
    abstract fun bindView(binding: ViewDataBinding, position: Int)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val binding: ViewDataBinding
        if (convertView == null) {
            binding = DataBindingUtil.inflate(inflater, getLayout(position), parent, false)
            view = binding.root
        } else {
            view = convertView
            binding = DataBindingUtil.getBinding(convertView)!!
        }
        bindView(binding, position)
        return view
    }

    init {
        inflater = LayoutInflater.from(context)
    }
}